zkevm-circuits/src/evm_circuit/execution/copy_code_to_memory.rs

```diff
@@ -0,0 +1,488 @@
+                let next_dst_addr = cb.query_cell();
+                let next_bytes_left = cb.query_cell();
+                let next_src_addr_end = cb.query_cell();
+                let next_code_hash = cb.query_word();
+
+                cb.require_equal(
+                    "next_src_addr == src_addr + copied_size",
+                    next_src_addr.expr(),
+                    src_addr.expr() + copied_size.clone(),
+                );
+                cb.require_equal(
+                    "dst_addr + copied_size == next_dst_addr",
+                    next_dst_addr.expr(),
+                    dst_addr.expr() + copied_size.clone(),
+                );
+                cb.require_equal(
+                    "next_bytes_left == bytes_left - copied_size",
+                    next_bytes_left.expr(),
+                    bytes_left.expr() - copied_size.clone(),
+                );
+                cb.require_equal(
+                    "next_src_addr_end == src_addr_end",
+                    next_src_addr_end.expr(),
+                    src_addr_end.expr(),
+                );
+                cb.require_equal(
+                    "next_code_hash == code_hash",
+                    next_code_hash.expr(),
+                    code_hash.expr(),
+                );
+            },
+        );
+
+        // Since this is an internal step for `CODECOPY` opcode, we only increment the
+        // RW counter. The program counter, stack pointer, and other fields do
+        // not change.
+        let step_state_transition = StepStateTransition {
+            rw_counter: Transition::Delta(cb.rw_counter_offset()),
+            ..Default::default()
+        };
+        cb.require_step_state_transition(step_state_transition);
+
+        Self {
+            src_addr,
+            dst_addr,
+            bytes_left,
+            src_addr_end,
+            code_hash,
+            is_codes,
+            buffer_reader,
+            finish_gadget,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction,
+        _call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        // Read the auxiliary data.
+        let (src_addr, dst_addr, bytes_left, src_addr_end, code, selectors) =
+            if let StepAuxiliaryData::CopyCodeToMemory {
+                src_addr,
+                dst_addr,
+                bytes_left,
+                src_addr_end,
+                code,
+                selectors,
+            } = step
+                .aux_data
+                .as_ref()
+                .expect("could not find aux_data for COPYCODETOMEMORY")
+            {
+                (
+                    src_addr,
+                    dst_addr,
+                    bytes_left,
+                    src_addr_end,
+                    code,
+                    selectors,
+                )
+            } else {
```

Comment:
not sure Line L190 expect statement can cover the `else` condition ?

Reply:
The L190 is to deal with auxiliary data being optional. So L190 `expect`s a `Some<T>` while the `else` handles the case where that auxiliary data is not `StepAuxiliaryData::CopyCodeToMemory` variant of the enum.